lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,75 @@
+module RSpec
+  module Matchers
+    # @api private
+    # Handles list of expected values when there is a need to render
+    # multiple diffs. Also can handle one value.
+    class ExpectedsForMultipleDiffs
+      # @private
+      # Default diff label when there is only one matcher in diff
+      # output
+      DEFAULT_DIFF_LABEL = "Diff:".freeze
+
+      # @private
+      # Maximum readable matcher description length
+      DESCRIPTION_MAX_LENGTH = 65
+
+      def initialize(expected_list)
+        @expected_list = expected_list
+      end
+
+      # @api private
+      # Wraps provided expected value in instance of
+      # ExpectedForMultipleDiffs. If provided value is already an
+      # ExpectedForMultipleDiffs then it just returns it.
+      # @param [Any] expected value to be wrapped
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.from(expected)
+        return expected if self === expected
+        new([[expected, DEFAULT_DIFF_LABEL]])
+      end
+
+      # @api private
+      # Wraps provided matcher list in instance of
+      # ExpectedForMultipleDiffs.
+      # @param [Array<Any>] matchers list of matchers to wrap
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.for_many_matchers(matchers)
+        new(matchers.map { |m| [m.expected, diff_label_for(m)] })
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    private
+
+      def self.diff_label_for(matcher)
+        "Diff for (#{truncated(matcher.description)}):"
+      end
+
+      def self.truncated(description)
+        return description if description.length < DESCRIPTION_MAX_LENGTH
+        description[0..DESCRIPTION_MAX_LENGTH] + "..."
+      end
```

Comment:
This isn't really a big deal (and certainly not a merge blocker), but if the description is too long it will actually wind up longer than `DESCRIPTION_MAX_LENGTH` (due to the addition of `...`).  What's slightly odd is that if a description is 66 or 67 characters, it'll get truncated to 65, and the `...` will get added...which will give us a _longer_ string than we had originally!

Seems like the string slice should stop 3 characters short to allow for `...`?

Also, I think we can use `<<` instead of `+` (one less string allocation).  The sliced string is safe to mutate since nothing else has a reference to it.